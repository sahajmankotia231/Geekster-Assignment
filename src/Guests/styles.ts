import styled from "styled-components";

import { WHITE, HEAVY_BLACK, BLUE } from "constants/colors";
import { ReactComponent as CloseSvg } from "assets/close.svg";

export const Container = styled.div`
  width: 100%;
  height: 100vh;
  background-color: ${WHITE};
  position: fixed;
  top: 0;
  left: 0;
  padding-bottom: 80px;
`;

export const Content = styled.div`
  height: calc(100% - 70px);
  overflow: auto;
  padding: 22px 16px 0;
`;

export const NavBar = styled.div`
  height: 64px;
  align-items: center;
  background-color: ${WHITE};
  box-shadow: 0px 0px 1px rgba(68, 80, 95, 0.08),
    0px 1px 4px rgba(68, 80, 95, 0.2);
  display: flex;
  position: relative;
  justify-content: center;
`;

export const NavBarTitle = styled.h3`
  color: ${HEAVY_BLACK};
  font-size: 16px;
  line-height: 20px;
  font-weight: 600;
`;

export const IconClose = styled(CloseSvg)`
  color: ${BLUE};
  fill: currentColor;
  position: absolute;
  top: 21px;
  left: 12px;
`;

export const Footer = styled.footer`
  height: 40px;
  position: fixed;
  left: 16px;
  right: 16px;
  bottom: 32px;
`;
