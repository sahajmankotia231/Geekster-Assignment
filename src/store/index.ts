import { createSlice, configureStore } from "@reduxjs/toolkit";

export type ChildType = {
  age: number | null;
};

export type RoomType = {
  maxNumberOfAdults: number;
  maxNumberOfChildren: number;
  maxOccupancy: number;
  adultsCounter: number;
  totalGuests: number;
  children: ChildType[];
};

export type StoreStateType = {
  maxNumberOfRooms: number;
  rooms: RoomType[];
};

const initialRoomsState = {
  maxNumberOfAdults: 5,
  maxNumberOfChildren: 3,
  maxOccupancy: 5,
  adultsCounter: 1,
  totalGuests: 1,
  children: [],
};

const initialState = {
  maxNumberOfRooms: 8,
  rooms: [initialRoomsState],
};

const counterSlice = createSlice({
  name: "rooms",
  initialState,
  reducers: {
    initializeStore: (state, action): void => {
      if (Boolean(action.payload)) {
        const splitedRooms = action.payload?.split("|") || [];
        const rooms = splitedRooms?.map(
          (room): RoomType => {
            const guests = room.split(":");
            const adultsCounter = Number(guests[0]);
            const children =
              guests[1]?.split(",")?.reduce((arr, age): ChildType[] => {
                arr.push({ age: Number(age) });
                return arr;
              }, []) || [];

            return {
              ...initialRoomsState,
              adultsCounter,
              children,
            };
          }
        );

        state.rooms = rooms;
      }
    },
    resetStore: (state): void => {
      state.rooms = [initialRoomsState];
    },
    addRoom: (state): void => {
      const { rooms, maxNumberOfRooms } = state;

      if (rooms.length < maxNumberOfRooms) {
        rooms.push(initialRoomsState);
      }
    },
    removeRoom: (state, action): void => {
      if (state.rooms.length > 1) {
        state.rooms.splice(action.payload, 1);
      }
    },
    addAdult: (state, action): void => {
      const room = state.rooms[action.payload];

      if (
        room.adultsCounter < room.maxNumberOfAdults &&
        room.totalGuests < room.maxOccupancy
      ) {
        room.adultsCounter += 1;
        room.totalGuests += 1;
      }
    },
    removeAdult: (state, action): void => {
      const room = state.rooms[action.payload];

      if (room.adultsCounter > 1) {
        room.adultsCounter -= 1;
        room.totalGuests -= 1;
      }
    },
    addChild: (state, action): void => {
      const room = state.rooms[action.payload];

      if (
        room.children.length < room.maxNumberOfChildren &&
        room.totalGuests < room.maxOccupancy
      ) {
        room.children.push({
          age: null,
        });
        room.totalGuests += 1;
      }
    },
    removeChild: (state, action): void => {
      const room = state.rooms[action.payload];
      const { children } = room;

      if (children.length) {
        children.splice(children.length - 1, 1);
        room.totalGuests -= 1;
      }
    },
    removeSpecificChild: (state, action): void => {
      const { childrenPosition, roomPosition } = action.payload;
      const room = state.rooms[roomPosition];
      const { children } = room;

      if (children.length) {
        children?.splice(childrenPosition, 1);
        room.totalGuests -= 1;
      }
    },
    changeChildAge: (state, action): void => {
      const { childrenPosition, roomPosition, age } = action.payload;
      const { children } = state.rooms[roomPosition];

      if (Boolean(age)) {
        children[childrenPosition].age = Number(age);
      }
    },
  },
});

export const {
  initializeStore,
  resetStore,
  addRoom,
  removeRoom,
  addAdult,
  removeAdult,
  addChild,
  removeChild,
  removeSpecificChild,
  changeChildAge,
} = counterSlice.actions;

const store = configureStore({
  reducer: counterSlice.reducer,
});

export default store;
