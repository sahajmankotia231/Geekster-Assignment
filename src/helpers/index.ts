import { RoomType } from "store";

export const serializeRooms = (rooms: Array<RoomType>): string => {
  const mappedRooms = rooms.map((room) => {
    let serializedRoom = String(room.adultsCounter);

    if (room.children.length) {
      const serializedChildren = room.children.reduce(
        (childStr, child, childIndex) => {
          if (childIndex + 1 <= room.children.length) {
            childStr += childIndex ? `,${child.age}` : child.age;
          }
          return childStr;
        },
        ""
      );

      serializedRoom += `:${serializedChildren}`;
    }

    return serializedRoom;
  }, []);

  const reducedRooms = mappedRooms.reduce((str, room, index) => {
    if (index + 1 <= mappedRooms.length) {
      str += index ? `|${room}` : room;
    }

    return str;
  }, "");

  return window.encodeURIComponent(reducedRooms);
};
