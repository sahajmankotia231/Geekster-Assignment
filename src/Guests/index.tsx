import React, { ReactElement } from "react";
import { useSelector, useDispatch } from "react-redux";

import Button from "components/Button";
import { serializeRooms } from "helpers";
import { StoreStateType, addRoom, resetStore } from "store";
import { ReactComponent as IconSearch } from "assets/search.svg";

import GuestsRoom from "./components/GuestsRoom";
import {
  Container,
  Content,
  NavBar,
  NavBarTitle,
  IconClose,
  Footer,
} from "./styles";

interface GuestsProps {
  onClose: () => void;
}

const Guests = ({ onClose }: GuestsProps): ReactElement => {
  const { rooms, maxNumberOfRooms } = useSelector(
    (state: StoreStateType) => state
  );
  const dispatch = useDispatch();
  const isAddRoomButtonDisabled = rooms.length === maxNumberOfRooms;
  const numberOfChildWithoutAge = rooms.reduce((acc, room) => {
    acc += room.children.filter((child) => child.age === null).length;
    return acc;
  }, 0);
  const isSearchButtonDisabled = Boolean(numberOfChildWithoutAge);

  const handleClose = (): void => {
    window.location.replace("/");
    dispatch(resetStore());
    onClose();
  };

  const handleSearch = (): void => {
    window.location.replace(
      window.location.origin + "?rooms=" + serializeRooms(rooms)
    );
    onClose();
  };

  const getSearchButtonText = (): string => {
    const roomsText = rooms.length > 1 ? " rooms" : " room";
    const totalGuests = rooms.reduce((acc, room) => {
      acc += room.totalGuests;
      return acc;
    }, 0);
    const guestsText = totalGuests > 1 ? " guests" : " guest";

    return `Search ${rooms.length + roomsText} â€¢ ${totalGuests + guestsText}`;
  };

  return (
    <Container>
      <NavBar>
        <IconClose onClick={handleClose} data-testid="icon-close" />
        <NavBarTitle>Who is staying</NavBarTitle>
      </NavBar>
      <Content>
        {rooms.map((room, index) => (
          <GuestsRoom key={`room-${index}`} position={index} />
        ))}
        <Button
          onClick={() => dispatch(addRoom())}
          theme="secondary"
          disabled={isAddRoomButtonDisabled}
        >
          + Add Room
        </Button>
      </Content>
      <Footer>
        <Button
          onClick={handleSearch}
          dataTestId="search-button"
          disabled={isSearchButtonDisabled}
        >
          <IconSearch />
          {getSearchButtonText()}
        </Button>
      </Footer>
    </Container>
  );
};

export default Guests;
