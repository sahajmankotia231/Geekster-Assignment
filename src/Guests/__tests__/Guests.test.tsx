import React from "react";
import { Provider } from "react-redux";
import { render, fireEvent, screen } from "@testing-library/react";

import { serializeRooms } from "helpers";
import store, { resetStore } from "store";

import Guests from "..";

jest.mock("helpers", () => ({
  serializeRooms: jest.fn(),
}));

const mockedReplace = jest.fn();

Object.defineProperty(window, "location", {
  value: {
    replace: mockedReplace,
  },
});

describe("<Guests />", () => {
  beforeEach(() => store.dispatch(resetStore()));

  it("should render properly", () => {
    const { container } = render(
      <Provider store={store}>
        <Guests onClose={() => {}} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should be able to close the component", () => {
    const mockedOnClose = jest.fn();
    render(
      <Provider store={store}>
        <Guests onClose={mockedOnClose} />
      </Provider>
    );

    fireEvent.click(screen.getByTestId("icon-close"));

    expect(mockedOnClose).toHaveBeenCalled();
    expect(mockedReplace).toHaveBeenCalled();
  });

  it("should be able to add more rooms", () => {
    render(
      <Provider store={store}>
        <Guests onClose={() => {}} />
      </Provider>
    );

    expect(screen.getAllByTestId("guests-room")).toHaveLength(1);

    fireEvent.click(screen.getByText("+ Add Room"));

    expect(screen.getAllByTestId("guests-room")).toHaveLength(2);
  });

  it("should be able to restore the store on close", () => {
    render(
      <Provider store={store}>
        <Guests onClose={() => {}} />
      </Provider>
    );

    fireEvent.click(screen.getByText("+ Add Room"));
    fireEvent.click(screen.getByText("+ Add Room"));
    expect(screen.getAllByTestId("guests-room")).toHaveLength(3);

    fireEvent.click(screen.getByTestId("icon-close"));

    expect(screen.getAllByTestId("guests-room")).toHaveLength(1);
  });

  describe("Search Button", () => {
    it("should have a text with total of rooms and guests", () => {
      render(
        <Provider store={store}>
          <Guests onClose={() => {}} />
        </Provider>
      );

      const searchButton = screen.getByTestId("search-button");

      expect(searchButton.textContent).toBe(
        "search.svgSearch 1 room • 1 guest"
      );

      fireEvent.click(screen.getByTestId("add-adult-0-button"));

      expect(searchButton.textContent).toBe(
        "search.svgSearch 1 room • 2 guests"
      );

      fireEvent.click(screen.getByText("+ Add Room"));
      fireEvent.click(screen.getByTestId("add-adult-1-button"));
      fireEvent.click(screen.getByTestId("add-child-0-button"));

      expect(searchButton.textContent).toBe(
        "search.svgSearch 2 rooms • 5 guests"
      );
    });

    it("should be disabled if the room has a child without age defined", () => {
      render(
        <Provider store={store}>
          <Guests onClose={() => {}} />
        </Provider>
      );

      const searchButton = screen.getByTestId("search-button");

      expect(searchButton).not.toHaveAttribute("disabled");

      fireEvent.click(screen.getByTestId("add-child-0-button"));
      expect(searchButton).toHaveAttribute("disabled");

      fireEvent.change(screen.getByTestId("child-0-0-select"), {
        target: { value: 0 },
      });
      expect(searchButton).not.toHaveAttribute("disabled");
    });

    it("should emit a search", () => {
      const mockedOnClose = jest.fn();

      render(
        <Provider store={store}>
          <Guests onClose={mockedOnClose} />
        </Provider>
      );

      const searchButton = screen.getByTestId("search-button");

      fireEvent.click(screen.getByTestId("add-child-0-button"));
      fireEvent.change(screen.getByTestId("child-0-0-select"), {
        target: { value: 0 },
      });
      fireEvent.click(searchButton);

      expect(serializeRooms).toHaveBeenCalledWith(store.getState().rooms);
      expect(mockedOnClose).toHaveBeenCalled();
      expect(mockedReplace).toHaveBeenCalled();
    });
  });
});
