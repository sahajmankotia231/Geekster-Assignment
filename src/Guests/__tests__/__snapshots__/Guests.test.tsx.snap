// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Guests /> should render properly 1`] = `
<div
  class="sc-kDThTU dLjDPv"
>
  <div
    class="sc-crHlIS grFNon"
  >
    <svg
      class="sc-bqiQRQ dEzelX"
      data-testid="icon-close"
    >
      close.svg
    </svg>
    <h3
      class="sc-egiSv kYnSGl"
    >
      Who is staying
    </h3>
  </div>
  <div
    class="sc-iqsfdx jiwjTi"
  >
    <div
      class="sc-gsDJrp dwyZHA"
      data-testid="guests-room"
    >
      <div
        class="sc-dkPtyc iQYZgR"
      >
        <h4
          class="sc-hKwCoD cLeDZh"
        >
          Room 
          1
        </h4>
      </div>
      <div
        class="sc-eCImvq guCNbL"
      >
        <span
          class="sc-jRQAMF bgpgfc"
        >
          Adults
        </span>
        <div
          class="sc-gKckTs jduWvm"
        >
          <button
            class="sc-bdvvaa fzcnyp"
            data-testid="remove-adult-0-button"
            disabled=""
          >
            <svg>
              minus.svg
            </svg>
          </button>
          <span
            class="sc-jRQAMF bgpgfc"
            data-testid="adults-counter"
          >
            1
          </span>
          <button
            class="sc-bdvvaa fzcnyp"
            data-testid="add-adult-0-button"
          >
            <svg>
              plus.svg
            </svg>
          </button>
        </div>
      </div>
      <div
        class="sc-eCImvq guCNbL"
      >
        <span
          class="sc-jRQAMF bgpgfc"
        >
          Children
        </span>
        <div
          class="sc-gKckTs jduWvm"
        >
          <button
            class="sc-bdvvaa fzcnyp"
            data-testid="remove-child-0-button"
            disabled=""
          >
            <svg>
              minus.svg
            </svg>
          </button>
          <span
            class="sc-jRQAMF bgpgfc"
            data-testid="children-counter"
          >
            0
          </span>
          <button
            class="sc-bdvvaa fzcnyp"
            data-testid="add-child-0-button"
          >
            <svg>
              plus.svg
            </svg>
          </button>
        </div>
      </div>
      <div
        class="sc-iCfLBT bZPXmC"
      />
    </div>
    <button
      class="sc-bdvvaa fzcnyp"
      data-testid="button"
    >
      + Add Room
    </button>
  </div>
  <footer
    class="sc-ksdxAp guitwb"
  >
    <button
      class="sc-bdvvaa tPxtg"
      data-testid="search-button"
    >
      <svg>
        search.svg
      </svg>
      Search 1 room â€¢ 1 guest
    </button>
  </footer>
</div>
`;
